File: Lecture 7 Practice/t3-t6-t8.c
#include <stdio.h>

int main (void)
{
  int n = 0;
  unsigned int result = 1;
  int i = 0; // Tmp Counter
  int r = 0; // Ret Value

  r = scanf ("%d", &n);
  if (r != 1) {
    puts ("Illegal input.");
    return 1;
  }
  if (n < 0) {
    puts ("Meet negative number, expect positive one. Exiting.");
    return 1;
  }
  if (n >= sizeof (unsigned) * 8) {
    printf ("Warning: N larger than or equal to %d may cause unpredictable output.\n", sizeof (unsigned) * 8);
  }

  // Using while per requirement.
  // i = n;
  // while (i) {
  //   result *= 2;
  //   --i;
  // }

  // Using do-while per requirement.
  // i = n;
  // if (i == 0) goto output;
  //
  // do {
  //   result *= 2;
  //   --i;
  // } while (i);

  // Using for per requirement.
  for (i = n; i; --i) {
    result *= 2;
  }

// output: // For do-while's 0 condition jump
  printf ("2^%d is %u\n", n, result);

  return 0;
}

File: Lecture 7 Practice/t3-t6-t8.go
package main

import (
  "fmt"
  "os"
  "bufio"
  "unsafe"
)

func main() {
  var n int
  var result uint = 1
  var i int   // Tmp Counter
  var e error // Ret Error

  stdin := bufio.NewReader(os.Stdin)

  _, e = fmt.Fscanf(stdin, "%d", &n)
  if e != nil {
    fmt.Println("Illegal input: ", e)
    return
  }
  if n < 0 {
    fmt.Println("Meet negative number, expect positive one. Exiting.")
    return
  }
  if uintptr(n) >= (unsafe.Sizeof(uint) * 8) {
    fmt.Printf("Warning: N larger than or equal to %d may cause unpredictable output.\n", unsafe.Sizeof(uint) * 8)
  }

  // Using while per requirement
  // i = n
  // for i != 0 {
  //   result *= 2
  //   i -= 1
  // }

  // There is no do-while in Go.

  // Using for per requirement.
  for i = n; i != 0; i -= 1 {
    result *= 2
  }

// output:
  fmt.Printf("2^%d is %d\n", n, result)

  stdin.Discard(stdin.Buffered())
}
